### ISSUE TITLE
Axios request fail to authenticate while using headers

### ISSUE THREAD

**Issue by iamrajkaran** [axios/axios#2245]
**Describe the issue**
Upgrading from 0.18 to 0.19 version is causing an issue in HTTP requests - GET, POST when sending headers like cookie. 

**Example Code**

```
  const requestOptions = {
    baseURL: 'http://localhost:3000',
    url: `${endPoint}${ids}`,
    method: 'GET',
    responseType: 'json',
    timeout: 100000,
    headers: {
      VerificationToken: userToken,
      Cookie: cookie,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
  };

  axios(requestOptions)
    .then((response) => {
      // Trigger in case, if it is not authenticated
      if (typeof response.data === 'string') {
       console.log('It is working here');
      }else {
       console.log('It should Work here');
      }
    })
    .catch((error) => {
    });
```

**Expected behavior, if applicable**
Api should return expected result as it was giving in ver 0.18.0`

**Environment:**
 - Axios Version [e.g. 0.19.0]
 - OS: [e.g. OSX Mojave]
 - Browser [e.g. Chrome [Electron] ]
 - Electron Version [e.g. 4.0.4]

**Additional context/Screenshots**
It is working flawless in ver 0.18.0, but not in ver 0.19.0 if we hit request (GET, POST) that requires headers.

**Comment by slim-hmidi** [comment]
Could you verify if you missed something in your code or it is due to your cookie or VerificationToken? Because I test with a simple example and it works normally:
index.js:
```
const axios = require('axios')

const requestOptions = {
    url: 'http://localhost:3000/products',
    method: 'POST',
    responseType: 'json',
    timeout: 100000,
    data: {
        productName: 'My Product'
    },
    headers: {
      'Content-Type': 'application/json',
    },
  };

  axios(requestOptions)
    .then((response) => {
      // Trigger in case, if it is not authenticated
      if (typeof response.data === 'string') {
       console.log('It is working here');
      }else {
       console.log('It should Work here');
      }
    })
    .catch((error) => {
        return error
    });
```
server.js:
```
const express = require('express');
const bodyParser = require('body-parser')

app = express();

app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.get('/products/:id', (req, res) => {
    const {id} = req.params;
    console.log(req.params)
    res.send(`Product ${id} was returned`)
})

app.post('/products', (req, res) => {
    const {productName} = req.body;
    console.log(productName)
    res.send(`Product ${productName} was created`)
})

app.listen(3000, () => {
    console.log('Server listens on 127.0.0.1:3000')
})
```
And I run the `server.js` and I execute the `index.js` and it works. I think that you missed something in your code.

**Comment by rajmonotype** [comment]
The solution proposed by you works, but it doesn't work if we send a cookie in headers.
Is sending cookie mechanism changed in 0.19.0?

**Comment by iamrajkaran** [comment]
@slim-hmidi 
The solution proposed by you works, but it doesn't work if we send a cookie in headers.
Is sending cookie mechanism changed in 0.19.0?

**Comment by slim-hmidi** [comment]
@iamrajkaran  I edited the answer.Could you check again?

**Comment by iamrajkaran** [comment]
@slim-hmidi I believe I did the same but doesn't work for me.

**Comment by slim-hmidi** [comment]
I execute the code in different OS and it works for me.
First I execute on Ubunut 18.04 and I rexecute on Windows 10 
Node: 10.13.0
axios: 0.19.0

**Comment by iamrajkaran** [comment]
It is working fine with 0.18.0 and there are no such changes on server and client side. But still, it is generating an error. 
Have you tried with a sending headers witrh cookie? I believe there might be an issue in receiving a cookie.

**Comment by slim-hmidi** [comment]
Could you provide me a reproductive example to test it on my pc?

**Comment by slim-hmidi** [comment]
In your code, you mention:
```
headers: { Cookie: cookie}
```
Could you give an example how you send the cookie. I tried this code:
```
const cookie = 'random cookie'
headers: {
Cookie: cookie
}
```
In the server, when you make `console.log(req.cookies)` you will get an empty object.

**Comment by LucaMozzo** [comment]
I tried both 0.18.0 and 0.19.0 and I get the same problem. I'm writing tests using a cookie-based authentication and the cookies are not set in the request object I get in the back-end.

```
let headers: any = {};
headers['Cookie'] = 'cookie value';
// in another function: 
const config: AxiosRequestConfig = {headers, withCredentials: true};
axios.get(url, config);
```
When I debug, I see the cookie in the headers, but not in the request cookies, and I'm sure about the back-end implementation as it is currently working with other clients

**Comment by slim-hmidi** [comment]
@LucaMozzo try :
```
headers['Cookie'] = 'cookie=value';

```

**Comment by LucaMozzo** [comment]
@slim-hmidi apologies, I might have oversimplified the example, I have 2 cookies in that format `name=urlencoded-value` separated by a semicolon

**Comment by slim-hmidi** [comment]
I don't know what to say about that, because I think the code that I written does not work in your case. Could you provide please a simple example that you wrote and mention your machine settings (OS, node, axios, etc...)

**Comment by iamrajkaran** [comment]
@LucaMozzo in fact in my case cookie is object and it is working fine in 0.18.0
my cookies is an object
// sample
{
0: 'string',
1: 'string2'
}

**Comment by slim-hmidi** [comment]
@LucaMozzo I think that sending cookies as an object is not supported and you should send them `cookieName=cookieValue; cookieName1=cookieValue1; cookieName2=cookieValue2`.
Did you try to send them like that?

**Comment by LucaMozzo** [comment]
@iamrajkaran I agree with @slim-hmidi , if I try to send an object it doesn't get deserialised not even in the headers. I managed to solve playing around with the header values and using v0.18.1

**Comment by feed4rz** [comment]
I am having the same problem.

On version 0.18.1 I was able to pass array of cookies `['a=b', 'c=d', ...]` to headers option:

```js
axios.get('https://google.com', {
  headers: {
    cookie: ['a=b', 'c=d']
  }
})
```

And it produced a valid cookie string in headers:
```curl
GET / HTTP/1.1\r\nAccept: application/json, text/plain, */*\r\ncookie: a=b; c=d\r\nUser-Agent: axios/0.18.1\r\nHost: google.com\r\nConnection: close\r\n\r\n
```

But since version 0.19.0 things have changed:
```curl
GET / HTTP/1.1\r\nAccept: application/json, text/plain, */*\r\ncookie: [object Object]\r\nUser-Agent: axios/0.19.0\r\nHost: google.com\r\nConnection: close\r\n\r\n
```

Array of cookies is not getting formatted anymore. **Is this a bug or a feature?**

**Comment by chinesedfan** [comment]
Closed due to stale. And feel free to open a new issue if someone met similar problem.

@feed4rz It should be related to #2567 and will be fixed in #2844.
